torch/torch_version.py

```diff
@@ -0,0 +1,68 @@
+from functools import total_ordering
+from typing import Iterable, Union
+
+from pkg_resources import packaging  # type: ignore[attr-defined]
+
+Version = packaging.version.Version
+InvalidVersion = packaging.version.InvalidVersion
+
+from .version import __version__ as internal_version
+
+
+@total_ordering
+class TorchVersion(str):
+    """A string with magic powers to compare to both Version and iterables!
+
+    Prior to 1.10.0 torch.__version__ was stored as a str and so many did
+    comparisons against torch.__version__ as if it were a str. In order to not
+    break them we have TorchVersion which masquerades as a str while also
+    having the ability to compare against both packaging.version.Version as
+    well as tuples of values, eg. (1, 2, 1)
+
+    Examples:
+        Comparing a TorchVersion object to a Version object
+            TorchVersion('1.10.0a') > Version('1.10.0a')
+
+        Comparing a TorchVersion object to a Tuple object
+            TorchVersion('1.10.0a') > (1, 2)    # 1.2
+            TorchVersion('1.10.0a') > (1, 2, 1) # 1.2.1
+
+        Comparing a TorchVersion object against a string
+            TorchVersion('1.10.0a') > '1.2'
+            TorchVersion('1.10.0a') > '1.2.1'
+    """
+    # fully qualified type names here to appease mypy
+    def _convert_to_version(self, inp: Union[packaging.version.Version, str, Iterable]) -> packaging.version.Version:
+        if isinstance(inp, Version):
+            return inp
+        elif isinstance(inp, str):
+            return Version(inp)
```

Comment:
I feel like this could potentially mask some useful information (s.a. "1.9.0+cu102"). Also, I think the packaging.Version returns `LegacyVersion` for cases like "torch-deploy-1.8"

Reply:
Actually for cases like `torch-deploy-1.8` it'd raise an `InvalidVersion` exception thus causing `TorchVersion` to fall back into regular string comparison.

```

In [3]: Version('torch-deploy-1.8')
---------------------------------------------------------------------------
InvalidVersion                            Traceback (most recent call last)
<ipython-input-3-c4aa93e1d2e3> in <module>
----> 1 Version('torch-deploy-1.8')

~/work/pytorch-stack/.venv/lib/python3.8/site-packages/pkg_resources/_vendor/packaging/version.py in __init__(self, version)
    275         match = self._regex.search(version)
    276         if not match:
--> 277             raise InvalidVersion("Invalid version: '{0}'".format(version))
    278 
    279         # Store the parsed out pieces of the version

InvalidVersion: Invalid version: 'torch-deploy-1.8'
```

For cases like `1.9.0+cu102` the information about the `+local` version will still remain in the `.local` function.

```
In [6]: v = Version('1.9.0+cu102')

In [7]: v.major
Out[7]: 1

In [8]: v.minor
Out[8]: 9

In [9]: v.micro
Out[9]: 0

In [10]: v.post

In [11]: v.local
Out[11]: 'cu102'
```

'+local' versioning is respected even when comparing against a string:
```
In [14]: from torch.torch_version import TorchVersion

In [15]: TorchVersion('1.9.0+cu102') == '1.9.0+cu102'
Out[15]: True
```

Reply:
Got it. There is another little snag -- do we want to compare the local versions? In my opinion, if the user has "1.9.0+cu101", it shouldn't matter if it is cuda or not, right? If there is a check such as `TorchVersion("1.9.0+cu102") == "1.9.0"`, I think it should return True. wdyt?